src/librepgp/stream-parse.cpp

```diff
@@ -2084,48 +2154,86 @@ init_encrypted_src(pgp_parse_handler_t *handler, pgp_source_t *src, pgp_source_t
+                 * session key.  This format helps reduce traffic analysis of messages. */
+                rnp_identifier_iterator_t it = NULL;
+                rnp_key_handle_t          handle = NULL;
+                const char *              identifier = NULL;
+
+                /* Iterate through the keys */
+                rnp_ffi_t ffi = (rnp_ffi_t) handler->key_provider->userdata;
+                if (rnp_identifier_iterator_create(ffi, &it, "keyid")) {
+                    return false;
                 }
-            } else {
-                decrypted_seckey = &seckey->pkt();
-            }
 
-            /* Try to initialize the decryption */
-            if (encrypted_try_key(param, &pubenc, decrypted_seckey, *handler->ctx->ctx)) {
-                have_key = true;
-                /* inform handler that we used this pubenc */
-                if (handler->on_decryption_start) {
-                    handler->on_decryption_start(&pubenc, NULL, handler->param);
-                }
-            }
+                while (!have_key && !rnp_identifier_iterator_next(it, &identifier)) {
+                    if (!identifier) {
+                        break;
+                    }
+                    if (rnp_locate_key(ffi, "keyid", identifier, &handle) || !handle) {
+                        RNP_LOG("Locating key %s failed", identifier);
+                        break;
+                    }
 
-            /* Destroy decrypted key */
-            if (seckey->encrypted()) {
-                delete decrypted_seckey;
-                decrypted_seckey = NULL;
-            }
+                    bool key_is_secret = false;
+                    rnp_key_have_secret(handle, &key_is_secret);
+                    bool key_can_decrypt = false;
+                    if (key_is_secret) {
+                        key_can_decrypt = handle->sec->can_encrypt();
+                    }
+                    rnp_key_handle_destroy(handle);
+                    // Try only secret, encrypt-capable keys
+                    if (!(key_is_secret && key_can_decrypt)) {
+                        continue;
+                    }
 
-            if (have_key) {
-                break;
+                    /* Parse keyid into binary form and put into pubenc object instead of zeros
+                     */
+                    assert(strlen(identifier) == PGP_KEY_ID_SIZE * 2);
+                    assert(pubenc.key_id.size() == PGP_KEY_ID_SIZE);
+                    if (rnp::hex_decode(identifier,
+                                        pubenc.key_id.data(),
+                                        pubenc.key_id.size()) != PGP_KEY_ID_SIZE) {
+                        assert(false);
+                        RNP_LOG("Internal error: failed to parse keyid %s", identifier);
+                        continue;
+                    }
+
+                    errcode = key_fetch_and_try(handler, param, pubenc, /*speculative=*/true);
+                    if (errcode == RNP_SUCCESS) {
```

Comment:
At some point we sticked to the shorter format - `if (!errcode) {...`

Reply:
ok